### PR TITLE
bunq header handling as enum (bunq/sdk_java#93) 

### PR DESCRIPTION
**Pull request by tubbynl**
i extracted magic constants for headers into a BunqHeader enum, added some junit tests on it

behaviour is as-is, but some tests are not really runnable locally

i'm using the bunq api currently in a spring-boot setup using rest-template, most of the gson/models i can use for deserializing the responses and in order to mimic the header handling i needed to copy some of the header names into my project.

i extracted that into an BunqHeader enum which you see here, if this get's into the sdk_java project i can remove this from my project :)

### CODE REVIEW

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
+package com.bunq.sdk.http;
+
+import java.util.Collection;
+import java.util.Map;
+
+public enum BunqHeader {
+    attachmentDescription("X-Bunq-Attachment-Description"),
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: make the enum CAP_SNAKE_CASE please see: https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+    BunqHeader(String header) {
+        this(header, null);
+    }
+
+    BunqHeader(String header, String defaultValue) {
+        this.header = header;
+        this.defaultValue = defaultValue;
+    }
+
+    public static BunqHeader parse(String value) {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: `parseHeaderOrNull` to prevent nullpointer.

**Comment by tubbynl** [comment]
need to look in to this, headers are probably not-yet as they were before, i'll extract a testcase of header content in the correct case

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+        for (BunqHeader header:values()) {
+            if (header.equals(value)) {
+                return header;
+            }
+        }
+
+        return null;
+    }
+
+    public String getHeader() {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: `getHeaderName` as this returns the name of the header right  ?

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        return header;
+    }
+
+    public String getDefaultValue() {
+        return defaultValue;
+    }
+
+    private String getOrDefault(String value) {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: `getHeaderValueOrDefault`

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+    public boolean equals(String header) {
+        return getHeader().equalsIgnoreCase(header);
+    }
+
+    public boolean isBunq() {
+        return getHeader().startsWith(PREFIX);
+    }
+
+    private String findKey(Collection<String> keys) {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: `..OrNull`

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+        return null;
+    }
+
+    public String getOrDefault(Map<String, String> headers) {
+        String key = findKey(headers.keySet());
+        if (key != null && headers.get(key) != null) {
+            return headers.get(key);
+        }
+        return getDefaultValue();
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: missing new line before return.

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+                return key;
+            }
+        }
+
+        return null;
+    }
+
+    public String getOrDefault(Map<String, String> headers) {
+        String key = findKey(headers.keySet());
+        if (key != null && headers.get(key) != null) {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: missing new line before if statement.

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+        for (String key:keys) {
+            if (this.equals(key)) {
+                return key;
+            }
+        }
+
+        return null;
+    }
+
+    public String getOrDefault(Map<String, String> headers) {
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: get what 🤔 I think this also returns value right ?

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 9]
```diff
@@ -0,0 +1,100 @@
+package com.bunq.sdk.http;
+
+import java.util.Collection;
+import java.util.Map;
+
+public enum BunqHeader {
+    attachmentDescription("X-Bunq-Attachment-Description"),
+    cacheControl("Cache-Control", "no-cache"),
+    contentType("Content-Type"),
+    clientAuthentication("X-Bunq-Client-Authentication"),
```
Comment (OGKevin) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: For better read and usability, perpend `BUNQ_`  to all the bunq headers. `isBunq()` should stay nevertheless.

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class BunqHeaderTest {
+    @Test
+    public void parse() {
+        // parse works case-insensitive
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: Remove this comment and put it in the test name. e.g. `parseCaseInsensitveTest`

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+        assertEquals(BunqHeader.clientResponseId, BunqHeader.parse("X-Bunq-Client-Response-Id"));
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: please extract this magic value in a constant.

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+        assertEquals(BunqHeader.clientResponseId, BunqHeader.parse("x-bunq-client-response-id"));
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: same

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+        assertTrue(BunqHeader.clientResponseId.isBunq());
+        assertTrue(BunqHeader.geolocation.isBunq());
+        assertTrue(BunqHeader.region.isBunq());
+        assertTrue(BunqHeader.serverSignature.isBunq());
+        assertTrue(BunqHeader.clientEncryptionKey.isBunq());
+        assertTrue(BunqHeader.clientRequestId.isBunq());
+    }
+
+    @Test
+    public void getOrDefault() {
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: get what or default ?

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 14]
```diff
@@ -0,0 +1,69 @@
...
+                h.getOrDefault(Collections.singletonMap("x-bunq-some-other-header", "test-id")));
+        assertEquals("Could not determine response id.",
+                h.getOrDefault(Collections.<String, String>emptyMap()));
+    }
+
+    @Test
+    public void addToMap() {
+        Map<String, String> headers = new HashMap<>();
+
+        //sut
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: This comment can go.

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+        BunqHeader.language.addTo(headers, null);
+        BunqHeader.geolocation.addTo(headers, null);
+        BunqHeader.userAgent.addTo(headers, "test-agent");
+
+        // verify
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: This one also

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 16]
```diff
@@ -0,0 +1,69 @@
...
+    }
+
+    @Test
+    public void addToMap() {
+        Map<String, String> headers = new HashMap<>();
+
+        //sut
+        BunqHeader.language.addTo(headers, null);
+        BunqHeader.geolocation.addTo(headers, null);
+        BunqHeader.userAgent.addTo(headers, "test-agent");
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: Please extract magic value in constant.

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+
+        // verify
+        assertEquals("en_US", headers.get("X-Bunq-Language"));
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: constant.

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+        assertEquals("0 0 0 0 000", headers.get("X-Bunq-Geolocation"));
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: Also constant.

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 19]
```diff
@@ -0,0 +1,69 @@
...
+        assertEquals("test-agent", headers.get("User-Agent"));
```
Comment (OGKevin) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: and constants.

**Comment by OGKevin** [comment]
@patrickdw1991 please 👀

File: src/test/java/com/bunq/sdk/model/generated/endpoint/AvatarTest.java [hunk 20]
```diff
@@ -53,11 +52,11 @@ public void createAvatarTest() {
...
-    customHeaders.put(ApiClient.HEADER_CONTENT_TYPE, CONTENT_TYPE);
+    HashMap<String, String> allCustomHeader = new HashMap<>();
 
-    return AttachmentPublic.create(customHeaders, file_contents).getValue();
-  }
+    BunqHeader.ATTACHMENT_DESCRIPTION.addTo(allCustomHeader, ATTACHMENT_PATH_IN);
+    BunqHeader.CONTENT_TYPE.addTo(allCustomHeader, CONTENT_TYPE);
 
+    return AttachmentPublic.create(allCustomHeader, file_contents).getValue();
+  }
```
Comment (patrickdw1991) [inline src/test/java/com/bunq/sdk/model/generated/endpoint/AvatarTest.java]: New line

File: src/test/java/com/bunq/sdk/http/BunqHeaderTest.java [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    Map<String, String> headers = new HashMap<>();
+
+    BunqHeader.LANGUAGE.addTo(headers, null);
+    BunqHeader.GEOLOCATION.addTo(headers, null);
+    BunqHeader.USER_AGENT.addTo(headers, HEADER_VALUE_USER_AGENT);
+
+    assertEquals(HEADER_VALUE_LANGUAGE, headers.get(BunqHeader.LANGUAGE.getHeaderName()));
+    assertEquals(HEADER_VALUE_GEOLOCATION, headers.get(BunqHeader.GEOLOCATION.getHeaderName()));
+    assertEquals(HEADER_VALUE_USER_AGENT, headers.get(BunqHeader.USER_AGENT.getHeaderName()));
+  }
```
Comment (patrickdw1991) [inline src/test/java/com/bunq/sdk/http/BunqHeaderTest.java]: New line

File: src/main/java/com/bunq/sdk/http/BunqRequestBuilder.java [hunk 22]
```diff
@@ -205,13 +213,7 @@ public BunqRequestBody getBody() {
...
-  public List<BunqBasicHeader> getAllHeaderAsList() {
+  public List<BunqBasicHeader> getAllHeader() {
     return this.allHeader;
   }
-
-  /**
-   */
-  public BunqBasicHeader[] getAllHeaderAsArray() {
-    return this.allHeader.toArray(new BunqBasicHeader[this.allHeader.size()]);
-  }
```
Comment (patrickdw1991) [inline src/main/java/com/bunq/sdk/http/BunqRequestBuilder.java]: New line

File: src/main/java/com/bunq/sdk/http/BunqHeader.java [hunk 23]
```diff
@@ -0,0 +1,102 @@
...
+
+    public String getHeaderValueOrDefault(Map<String, String> headers) {
+        String key = findKeyOrNull(headers.keySet());
+
+        if (key != null && headers.get(key) != null) {
+            return headers.get(key);
+        }
+
+        return getDefaultValue();
+    }
```
Comment (patrickdw1991) [inline src/main/java/com/bunq/sdk/http/BunqHeader.java]: New line

**Pull request merged**